### PR TITLE
Scheduling and Approvals UI tests

### PR DESCRIPTION
**Pull request by hellerve**
### Relates to #125 

This PR adds view/UI test cases for scheduling & approvals (and is thus basically a sequel to #879).

As of now, all job views are untested. We still have to figure out how to approach this: should all job views be tested in the context of this PR?

Cheers

### CODE REVIEW

File: nautobot/extras/tests/test_views.py [hunk 1]
```diff
@@ -909,3 +912,104 @@ def test_create_object_with_constrained_permission(self):
         for cf in CustomField.objects.all():
             cf.delete()
         super().test_create_object_with_constrained_permission()
+
+
+class ScheduledJobTestCase(
+    ViewTestCases.ListObjectsViewTestCase,
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_views.py:918]: Since we have `ScheduledJobView` and `ScheduledJobDeleteView` classes, should we also include `GetObjectViewTestCase` and `DeleteObjectViewTestCase` here?

File: nautobot/extras/tests/test_views.py [hunk 2]
```diff
@@ -909,3 +912,104 @@ def test_create_object_with_constrained_permission(self):
...
+        )
+        ScheduledJob.objects.create(
+            name="test3",
+            task="nautobot.extras.jobs.scheduled_job_handler",
+            job_class="-",
+            interval=JobExecutionType.TYPE_IMMEDIATELY,
+            user=user,
+            start_time=datetime.now(),
+        )
+
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_views.py:949]: Might be good to include a custom test function here to verify that only jobs that are enabled and approved appear in this listing?

File: nautobot/extras/tests/test_views.py [hunk 3]
```diff
@@ -909,3 +912,104 @@ def test_create_object_with_constrained_permission(self):
...
+        ScheduledJob.objects.create(
+            name="test3",
+            task="nautobot.extras.jobs.scheduled_job_handler",
+            job_class="-",
+            interval=JobExecutionType.TYPE_IMMEDIATELY,
+            user=user,
+            approval_required=True,
+            start_time=datetime.now(),
+        )
+
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_views.py:991]: Might be good to add a custom test function here to verify that only jobs needing approval are included in the listing?

File: nautobot/extras/tests/test_views.py [hunk 4]
```diff
@@ -909,3 +912,104 @@ def test_create_object_with_constrained_permission(self):
...
+
+class JobResultTestCase(
+    ViewTestCases.ListObjectsViewTestCase,
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_views.py:994]: Should be able to include `GetObjectViewTestCase`, `DeleteObjectViewTestCase` and `BulkDeleteObjectsViewTestCase` here as well?

File: nautobot/extras/views.py [hunk 5]
```diff
@@ -1001,16 +1009,18 @@ class JobResultBulkDeleteView(generic.BulkDeleteView):
     table = tables.JobResultTable
 
 
-class JobResultView(ContentTypePermissionRequiredMixin, View):
+class JobResultView(ObjectPermissionRequiredMixin, View):
     """
     Display a JobResult and its data.
     """
 
+    queryset = JobResult.objects
```
Comment (glennmatthews) [inline nautobot/extras/views.py:1017]: Perhaps?
```suggestion
    queryset = JobResult.objects.all()
```

File: nautobot/extras/views.py [hunk 6]
```diff
@@ -1001,16 +1009,18 @@ class JobResultBulkDeleteView(generic.BulkDeleteView):
     table = tables.JobResultTable
 
 
-class JobResultView(ContentTypePermissionRequiredMixin, View):
+class JobResultView(ObjectPermissionRequiredMixin, View):
```
Comment (glennmatthews) [inline nautobot/extras/views.py:1012]: I wonder if this should actually just become a `generic.ObjectView`?

File: nautobot/extras/tests/test_views.py [hunk 7]
```diff
@@ -925,28 +956,45 @@ def setUpTestData(cls):
...
             user=user,
             start_time=datetime.now(),
         )
 
+    def test_only_enabled_is_listed(self):
+        self.add_permissions("extras.view_scheduledjob")
+
+        response = self.client.get(self._get_url("list"))
+        self.assertHttpStatus(response, 200)
+        self.assertNotIn("test4", extract_page_body(response.content.decode(response.charset)))
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_views.py:996]: Should we also assert that test1 through test3 *are* in the page body?

File: nautobot/extras/tests/test_views.py [hunk 8]
```diff
@@ -988,28 +1036,234 @@ def setUpTestData(cls):
...
+            approval_required=False,
+            start_time=datetime.now(),
+        )
+
+    def test_only_approvable_is_listed(self):
+        self.add_permissions("extras.view_scheduledjob")
+
+        response = self.client.get(self._get_url("list"))
+        self.assertHttpStatus(response, 200)
+        self.assertNotIn("test4", extract_page_body(response.content.decode(response.charset)))
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_views.py:1054]: Same question/suggestion here.

File: nautobot/extras/tests/test_views.py [hunk 9]
```diff
@@ -988,28 +1036,234 @@ def setUpTestData(cls):
...
         JobResult.objects.create(
-            name="test3",
+            name="local/test_view/TestJob3",
             job_id=uuid.uuid4(),
             obj_type=obj_type,
         )
+
+
+class JobTestCase(
+    TestCase,
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_views.py:1087]: Perhaps add a comment or docstring explaining why this isn't using any of the generic base classes? (Because Job isn't a proper database model, right?)

File: nautobot/extras/tests/test_views.py [hunk 10]
```diff
@@ -988,28 +1036,234 @@ def setUpTestData(cls):
...
+    def test_list_without_permission(self):
+        self.assertHttpStatus(self.client.get(reverse("extras:job_list")), 403)
+
+    @override_settings(EXEMPT_VIEW_PERMISSIONS=["*"])
+    def test_list(self):
+        response = self.client.get(reverse("extras:job_list"))
+        self.assertHttpStatus(response, 200)
+
+        response_body = extract_page_body(response.content.decode(response.charset))
+        self.assertIn("DummyJob", response_body)
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_views.py:1099]: Does this test need to be skipped if the dummy plugin isn't installed?

I'm also surprised that this one doesn't need `JOBS_ROOT=DUMMY_JOBS` - what am I missing here?

File: nautobot/extras/tests/test_views.py [hunk 11]
```diff
@@ -988,28 +1036,234 @@ def setUpTestData(cls):
...
+    def test_run_now_no_worker(self):
+        self.add_permissions("extras.run_job")
+
+        data = {
+            "_schedule_type": "immediately",
+        }
+
+        response = self.client.post(reverse("extras:job", kwargs={"class_path": "local/test_pass/TestPass"}), data)
+
+        self.assertHttpStatus(response, 200)
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_views.py:1139]: Should we check for a message about the lack of a worker in this case?

File: nautobot/extras/tests/test_views.py [hunk 12]
```diff
@@ -988,28 +1036,234 @@ def setUpTestData(cls):
...
+        self.add_permissions("extras.run_job")
+
+        response = self.client.post(reverse("extras:job", kwargs={"class_path": "local/test_pass/TestPass"}))
+
+        self.assertHttpStatus(response, 200)
+        errors = extract_form_failures(response.content.decode(response.charset))
+        self.assertEqual(errors, ["_schedule_type: This field is required."])
+
+    @override_settings(EXEMPT_VIEW_PERMISSIONS=["*"], JOBS_ROOT=DUMMY_JOBS)
+    def test_run_now_no_worker(self):
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_views.py:1130]: Will this fail if there *is* a worker running (as is often the case when I'm developing locally)? Do we need to mock.patch the get_worker_count to guarantee that it returns 0 for this test?

File: nautobot/extras/tests/test_views.py [hunk 13]
```diff
@@ -988,28 +1036,234 @@ def setUpTestData(cls):
...
+        self.add_permissions("extras.run_job")
+
+        data = {
+            "_schedule_type": "immediately",
+        }
+
+        patched.return_value = 1
+        response = self.client.post(reverse("extras:job", kwargs={"class_path": "local/test_pass/TestPass"}), data)
+
+        self.assertHttpStatus(response, 302)
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_views.py:1153]: Should we assert as to where it redirects to? (`self.assertRedirects(...)`)

File: nautobot/extras/tests/test_views.py [hunk 14]
```diff
@@ -988,28 +1036,234 @@ def setUpTestData(cls):
...
+    def test_run_now(self, patched):
+        self.add_permissions("extras.run_job")
+
+        data = {
+            "_schedule_type": "immediately",
+        }
+
+        patched.return_value = 1
+        response = self.client.post(reverse("extras:job", kwargs={"class_path": "local/test_pass/TestPass"}), data)
+
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_views.py:1152]: Maybe we should mock.patch JobResult.enqueue_job (or equivalent) and assert that it was called as expected?

File: nautobot/extras/tests/test_views.py [hunk 15]
```diff
@@ -988,28 +1036,234 @@ def setUpTestData(cls):
...
+            "_schedule_type": "immediately",
+            "var": "12",
+        }
+
+        patched.return_value = 1
+        response = self.client.post(
+            reverse("extras:job", kwargs={"class_path": "local/test_required_args/TestRequired"}), data
+        )
+
+        self.assertHttpStatus(response, 302)
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_views.py:1186]: Same comments here about assertRedirects and checking that enqueue_job was called correctly.

File: nautobot/extras/tests/test_views.py [hunk 16]
```diff
@@ -988,28 +1036,234 @@ def setUpTestData(cls):
...
+        data = {
+            "_schedule_type": "future",
+            "_schedule_name": "test",
+            "_schedule_start_time": str(start_time),
+        }
+
+        patched.return_value = 1
+        response = self.client.post(reverse("extras:job", kwargs={"class_path": "local/test_pass/TestPass"}), data)
+
+        self.assertHttpStatus(response, 302)
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_views.py:1264]: assertRedirects - this will be a different redirect than the run_immediately tests right?

File: nautobot/extras/views.py [hunk 17]
```diff
@@ -20,7 +20,7 @@
 from nautobot.core.views import generic
 from nautobot.dcim.models import Device
 from nautobot.dcim.tables import DeviceTable
-from nautobot.extras.utils import get_worker_count
+from nautobot.extras import utils as extras_utils
```
Comment (glennmatthews) [inline nautobot/extras/views.py:23]: Why this change? Looks like you're still only using get_worker_count.

File: nautobot/extras/tests/test_views.py (hunk 7 above)
Comment (hellerve) [reply nautobot/extras/tests/test_views.py:996]: AFAIK this is already tested by the base class, right?

File: nautobot/extras/views.py (hunk 17 above)
Comment (hellerve) [reply nautobot/extras/views.py:23]: Yes, but to be able to patch it I cannot import it directly, we have to import the module. Otherwise the patch will not be picked up.

Comment (glennmatthews) [reply nautobot/extras/views.py:23]: Does it work if you import it directly and change the patch to `patch(nautobot.extras.views.get_worker_count)`?

File: nautobot/extras/tests/test_views.py (hunk 7 above)
Comment (glennmatthews) [reply nautobot/extras/tests/test_views.py:996]: True, to a limited extent - `test_list_objects_with_permission` doesn't, but `test_list_objects_with_constrained_permission` does. Good enough I guess.

File: nautobot/extras/views.py (hunk 17 above)
Comment (hellerve) [reply nautobot/extras/views.py]: Good call, I changed it to that!

File: nautobot/extras/tests/test_views.py (hunk 14 above)
Comment (hellerve) [reply nautobot/extras/tests/test_views.py:1161]: Patching `enqueue_job` would be a little problematic, since it’s responsible for creating a `JobResult`. But since the `JobResult` is created and we need to ensure that for the redirect, that might be just as good?

File: nautobot/extras/tests/test_views.py (hunk 10 above)
Comment (hellerve) [reply nautobot/extras/tests/test_views.py:1106]: Good question! The test seems to work for me; I think the dummy plugin is always installed during testing (though I don’t know where that happens)?

File: nautobot/extras/tests/test_views.py (hunk 15 above)
Comment (hellerve) [reply nautobot/extras/tests/test_views.py]: I’ve changed it to `assertRedirect` here as well.

File: nautobot/extras/tests/test_views.py (hunk 10 above)
Comment (glennmatthews) [reply nautobot/extras/tests/test_views.py:1106]: Really our tests should always be run with the dummy plugin installed since we would have lower test coverage without it - but there is an existing pattern for skipping tests just in case it somehow isn't installed - see `nautobot/extras/tests/test_plugins.py` and `test_api.py` for examples. 

To your question, `nautobot/core/tests/nautobot_config.py` adds "dummy_plugin" to `PLUGINS`, and the plugin is installed by Poetry since it's listed as a dev-dependency in `pyproject.toml` - so these days it *should* always be available for tests, I think.

File: nautobot/extras/tests/test_views.py (hunk 14 above)
Comment (glennmatthews) [reply nautobot/extras/tests/test_views.py:1163]: I think it's possible to pass through from a patch to the underlying method to allow for normal operation (using `mock.DEFAULT`, IIRC) but I agree that just checking that the JobResult is created and the redirect is correct should suffice.

File: nautobot/extras/tests/test_views.py [hunk 18]
```diff
@@ -1,13 +1,20 @@
+from datetime import datetime, timedelta
+import os.path
 import urllib.parse
 import uuid
 
+from django.conf import settings
 from django.contrib.auth import get_user_model
 from django.contrib.contenttypes.models import ContentType
+from django.test import override_settings, SimpleTestCase
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_views.py:9]: Based on the Travis failure, I think you need a `from django.http import Http404` somewhere in here.

**Review by glennmatthews — APPROVED** [review]

File: nautobot/extras/tests/test_views.py (hunk 14 above)
Comment (hellerve) [reply nautobot/extras/tests/test_views.py:1163]: Ah, thanks for the pointer! I’ll call it okay for now then :)

File: nautobot/extras/tests/test_views.py (hunk 18 above)
Comment (hellerve) [reply nautobot/extras/tests/test_views.py:10]: Ah, good catch, thanks!

File: nautobot/extras/tests/test_views.py (hunk 10 above)
Comment (hellerve) [reply nautobot/extras/tests/test_views.py:1107]: Thanks for the context! I changed this to use one of the jobs from `dummy_jobs` instead, such that we don’t rely on the plugin being present where we don’t have to :)

File: nautobot/extras/tests/test_views.py [hunk 19]
```diff
@@ -909,3 +923,361 @@ def test_create_object_with_constrained_permission(self):
...
+
+    @override_settings(EXEMPT_VIEW_PERMISSIONS=[], JOBS_ROOT=DUMMY_JOBS)
+    def test_get_without_permission(self):
+        response = self.client.get(reverse("extras:job", kwargs={"class_path": "local/test_pass/TestPass"}))
+        self.assertHttpStatus(response, 403)
+
+    @override_settings(EXEMPT_VIEW_PERMISSIONS=["*"], JOBS_ROOT=DUMMY_JOBS)
+    def test_get(self):
+        response = self.client.get(reverse("extras:job", kwargs={"class_path": "local/test_pass/TestPass"}))
+        self.assertHttpStatus(response, 200)
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_views.py:1117]: Looks like this is failing in Travis - for some reason we're getting a 404 here instead of a 200.

File: nautobot/extras/tests/test_views.py [hunk 20]
```diff
@@ -909,3 +923,361 @@ def test_create_object_with_constrained_permission(self):
...
+        if class_path.startswith("local/test_view"):
+            return self.TestJob
+        raise Http404
+
+    def setUp(self):
+        super().setUp()
+
+        # Monkey-patch the viewsets' _get_job methods to return our test class above
+        JobView._get_job = self.get_test_job_class
+        ScheduledJobView._get_job = self.get_test_job_class
```
Comment (jathanism) [inline nautobot/extras/tests/test_views.py:943]: Not that this is a huge issue, but is this necessary as part of every test method or could this be sufficient at the test case (class) level using `setUpClass()` instead? Just thinking about test efficiency.

**Review by jathanism — APPROVED** [review]
I have no major concerns that Glenn didn't already catch. Thank you for being so thorough!

File: nautobot/extras/tests/test_views.py [hunk 21]
```diff
@@ -908,3 +922,370 @@ def setUpTestData(cls):
             "http_method": "POST",
             "http_content_type": "application/json",
         }
+
+
+@override_settings(JOBS_ROOT=THIS_DIRECTORY)
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_views.py:927]: Now that #869 is merged, would you mind moving the block of Jobs tests up in this file so that they fall between `GraphQLQueriesTestCase` and `ObjectChangeTestCase` to keep the ordering consistent? (Note that I'm not asking you to separate this group of related tests - they can stay grouped together in that new location even though it's not "quite" alphabetical order overall.

File: nautobot/extras/tests/test_views.py [hunk 22]
```diff
@@ -908,3 +922,370 @@ def setUpTestData(cls):
...
+        )
+        ScheduledJob.objects.create(
+            name="test3",
+            task="nautobot.extras.jobs.scheduled_job_handler",
+            job_class="local/test_views/TestJob",
+            interval=JobExecutionType.TYPE_IMMEDIATELY,
+            user=user,
+            start_time=datetime.now(),
+        )
+        # this should not appear, since it’s not enabled
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_views.py:989]: I think the intermittent failure in CI is because `test_list_objects_with_constrained_permission()` just grabs the last two objects in `self._get_queryset()`, so unless we have a defined ordering for this queryset, it's going to get two more-or-less at random, which might or might not include this object.

I think you can fix it by *only* creating this object inside the `test_only_enabled_is_listed` test case, instead of class-wide here in `setUpTestData`.

Comment (hellerve) [reply nautobot/extras/tests/test_views.py]: Tried it out, thanks for the tip!

File: nautobot/extras/tests/test_views.py (hunk 21 above)
Comment (hellerve) [reply nautobot/extras/tests/test_views.py]: Done!

**Review by glennmatthews — APPROVED** [review]

File: nautobot/extras/tests/test_views.py (hunk 22 above)
Comment (hellerve) [reply nautobot/extras/tests/test_views.py]: That seems to have worked, yay!
